### PR TITLE
Refactor external Go API

### LINKED ISSUE #115: Clean up exported API
There are a lot of layers.  It's not clear when external callers should use a particular layer, and some of the layers may not be useful at all.

For example, in the versioned `config` packages, `TranslateBytes` takes a byte slice and returns a byte slice.  If the caller wants to pass in parsed FCC structs, or receive parsed Ignition structs, they might decide to call `Translate` instead.  But the only way to get validation (other than copypasting a bunch of boilerplate) is to use `TranslateBytes`.

We should document the intended use of the API.  Which functions are intended for external use?  Should users ever call directly into `base`?

### PR DESCRIPTION
**Pull request by bgilbert**
The external Go API is confusing and inconsistent.  It has multiple API layers, with unclear distinctions between them, and with no single API layer providing a complete set of functionality.  Clean it up.

- Instead of implementing a `config` package as the combination of one `base` package and one `distro` package, implement it as a subclass of either a `base` package or another `config` package.  Implement distro-specific sugar directly in the `config` package.  Drop the `distro` packages and disrecommend that client code directly use `base` packages.
- Instead of validating configs only in the API functions that take and emit `[]byte`, perform as much validation as possible in the lower-level functions that take and emit config structs, and update their return types to be more useful to client code.  Add `Unvalidated` variants of these lower-level functions for use by subclasses.
- Use consistent naming for external API functions.
- Rename translation option structs for clarity.
- Consolidate error definitions in `config/common`.
- Allow out-of-tree code to register a variant/vendor pair to be recognized by `config.TranslateBytes()`.
- Move utility code only needed by out-of-tree config implementations into a separate package.

Fixes #115.

cc @yuqi-zhang @LorbusChris

### CODE REVIEW

**Comment by LorbusChris** [comment]
This is amazing work, thank you @bgilbert!
/approve

File: config/common/errors.go [hunk 1]
```diff
@@ -19,6 +19,24 @@ import (
 )
 
 var (
+	// common field parsing
+	ErrNoVariant      = errors.New("Error parsing variant. Variant must be specified")
+	ErrInvalidVersion = errors.New("Error parsing version. Version must be a valid semver")
```
Comment (arithx) [inline config/common/errors.go:24]: I know these were previously capitalized but should we drop the casing?

**Review by arithx — APPROVED** [review]
Code looks great! Thanks for breaking up the commits like that; made the entire refactor very easy to review.

File: base/v0_1/translate.go [hunk 2]
```diff
@@ -30,13 +30,18 @@ import (
...
 	tr.AddCustomTranslator(translateFile)
 	tr.AddCustomTranslator(translateDirectory)
 	tr.AddCustomTranslator(translateLink)
-	translations, report := tr.Translate(&c, &ret)
-	return ret, translations, report
+
+	tm, r := translate.Prefixed(tr, "ignition", &c.Ignition, &ret.Ignition)
+	translate.MergeP(tr, tm, &r, "passwd", &c.Passwd, &ret.Passwd)
+	translate.MergeP(tr, tm, &r, "storage", &c.Storage, &ret.Storage)
+	translate.MergeP(tr, tm, &r, "systemd", &c.Systemd, &ret.Systemd)
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply base/v0_1/translate.go:43]: Yeah.  Ignition does something similar, though, for the same reason.

**Comment by bgilbert** [comment]
Thanks for the review!

**Pull request merged**